src/leap/mail/mail.py

```diff
@@ -0,0 +1,697 @@
+        self.mbox_indexer = mbox_indexer
+        self.mbox_wrapper = mbox_wrapper
+
+        # TODO --- review this count shit. I think it's better to patch server
+        # to accept deferreds.
+        # TODO need to initialize count here because imap server does not
+        # expect a defered for the count. caller should return the deferred for
+        # prime_count (ie, initialize) when returning the collection
+        # TODO should increment and decrement when adding/deleting.
+        # TODO recent count should also be static.
+
+        if not count:
+            count = 0
+        self._count = count
+
+    def is_mailbox_collection(self):
+        """
+        Return True if this collection represents a Mailbox.
+        :rtype: bool
+        """
+        return bool(self.mbox_wrapper)
+
+    @property
+    def mbox_name(self):
+        wrapper = getattr(self, "mbox_wrapper", None)
+        if not wrapper:
+            return None
+        return wrapper.mbox
+
+    @property
+    def mbox_uuid(self):
+        wrapper = getattr(self, "mbox_wrapper", None)
+        if not wrapper:
+            return None
+        return wrapper.uuid
+
+    def get_mbox_attr(self, attr):
+        return getattr(self.mbox_wrapper, attr)
+
+    def set_mbox_attr(self, attr, value):
+        setattr(self.mbox_wrapper, attr, value)
+        return self.mbox_wrapper.update(self.store)
+
+    # Get messages
+
+    def get_message_by_content_hash(self, chash, get_cdocs=False):
+        """
+        Retrieve a message by its content hash.
+        :rtype: Deferred
+        """
+
+        if not self.is_mailbox_collection():
+            # instead of getting the metamsg by chash, query by (meta) index
+            # or use the internal collection of pointers-to-docs.
+            raise NotImplementedError()
+
+        metamsg_id = _get_mdoc_id(self.mbox_name, chash)
+
+        return self.adaptor.get_msg_from_mdoc_id(
+            self.messageklass, self.store,
+            metamsg_id, get_cdocs=get_cdocs)
+
+    def get_message_by_uid(self, uid, absolute=True, get_cdocs=False):
+        """
+        Retrieve a message by its Unique Identifier.
+
+        If this is a Mailbox collection, that is the message UID, unique for a
+        given mailbox, or a relative sequence number depending on the absolute
+        flag. For now, only absolute identifiers are supported.
+        :rtype: Deferred
+        """
+        if not absolute:
+            raise NotImplementedError("Does not support relative ids yet")
+
+        def get_msg_from_mdoc_id(doc_id):
+            if doc_id is None:
+                return None
+            return self.adaptor.get_msg_from_mdoc_id(
+                self.messageklass, self.store,
+                doc_id, uid=uid, get_cdocs=get_cdocs)
+
+        d = self.mbox_indexer.get_doc_id_from_uid(self.mbox_uuid, uid)
+        d.addCallback(get_msg_from_mdoc_id)
+        return d
+
+    # TODO deprecate ??? ---
+    def _prime_count(self):
```

Comment:
removed. this is part of a series of cases in which I've got things that *are* deferreds, but the server implementation expects inmediate values. I've considered patching the server to use maybeDeferred, but maybe it's easier to wait for intializations (in our __init__ antipattern, or in a factory function as exarkun recommends in his post) and then just handle the values.

I think I'll tackle the several count_ methods right after we merge this onto async-api, since it's needed for the notifications to the MUA to work properly.